wordpress-mobile/WordPress-Android
# WP Stories: story composer ViewModel part4

Builds on top of #12504 

Comes from #12493 (see [comments](https://github.com/wordpress-mobile/WordPress-Android/pull/12493#issuecomment-663472463))


### Things done in this PR
- created a new StoryEditorMedia class as a starting point for a proper VM
- moved out a an enum and a listener we needed on both EditorMedia and StoryEditorMedia: `EditorMediaListener` interface and `AddExistingMediaSource` enum

removed:
- `refreshBlogMedia`, didn't use it
- `retryFailedMediaAsync`: didn't use it but maybe in the future
+ `purgeMediaToPostAssociationsIfNotInPostAnymoreAsync`: I think we'll need this one as soon as we allow editing (very soon, before launch). Took it away for now to leave things in a clean state, will revisit soon
- `reattachUploadingMediaForAztec`, we don't have Aztec.
- `definitelyDeleteBackspaceDeletedMediaItemsAsync` we don't need this either
- `updateDeletedMediaItemIds`: removed
+ `onMediaDeleted`: we'll want to use this when editing posts as well. Deleting for now.
- `onMediaUploadError`: didn't use it, we don't show upload errors on the UI.
- `updateMediaUploadStateBlocking` (we don't upload with the UI on the front)
- `cancelMediaUploadAsync` (we don't upload with the UI on the front)
- `addGifMediaToPostAsync` (we don't use gifs for now)
- `addFreshlyTakenVideoToEditor` (we add video differently because the StoryComposer has its own camera capture mode)
- `addExistingMediaToEditorAsync` variants we didn't use (removed)

#### Some notes / questions
- `AddMediaToStoryPostUiState`. This class is a copy & paste of `AddMediaToPostUiState` in `EditorMedia`. Wonder how to make this better; I didn't feel it was OK to reuse the same class. Probably not a big deal, but copy/paste leaves me with a bitter taste 🤷‍♂️ 


To test:
- Make sure EditPostActivity media handling works well (add media from different sources, reattach, cancel, etc)
- Make sure StoryComposerActivity works well (add media from different sources, save, publish). 

PR submission checklist:
